Add unit conversion, variable load limit to RAQMS

mbruckner-work opened the pull request:
Additions to RAQMS reader:

1) Convert gases to ppbv ; resolves #134 
2) option for loading a small set of variables (consistent with _mm readers)

Before merging plan to add renaming or calculation for temperature_k and dz_m (needed for some aircraft and satellite pairings). These updates will be post-tutorial.

Commit:
add unit conversion for gases

Commit:
add variable limit option to RAQMS

Commit:
ensure raqms reader update included

Commit:
Merge branch 'develop' of https://github.com/mbruckner-work/monetio into develop

Something local didn't line up

mbruckner-work committed:
Merge pull request #1 from noaa-oar-arl/develop

Update from ARL develop branch

mbruckner-work committed:
possible fix for tropopause index issue in OFFL files

mbruckner-work committed:
Merge branch 'noaa-oar-arl:stable' into raqms_update

mbruckner-work committed:
add temperature

Needed to calculate temperature from potential temperature

zmoon commented on monetio/models/raqms.py:
```diff
@@ -30,17 +32,37 @@ def open_dataset(fname, *, surf_only=False):
     fname : str or list of str
         Files to be opened, expressed as a glob string or list of string paths.
-
+    convert_to_ppb : boolean
```
```suggestion
    convert_to_ppb : bool
```

zmoon commented on monetio/models/raqms.py:
```diff
@@ -30,17 +32,37 @@ def open_dataset(fname, *, surf_only=False):
+    if convert_to_ppb:
+        for i in ds.variables:
+            if "units" in ds[i].attrs:
+                if "ppv" in ds[i].attrs["units"]:
```
```suggestion
                if ds[i].attrs["units"].startswith("ppv"):
```
a little more specific? (assuming I'm correct it would be in the start) or if it's always just 'ppv' you could use `== "ppv"` instead.

mbruckner-work replied:
I was following the formatting used by the other readers. Switching to `== "ppv"`.

zmoon commented on monetio/models/raqms.py:
```diff
@@ -9,14 +9,16 @@
     fname : str
         File to be opened.
-
+    convert_to_ppb : boolean
```
```suggestion
    convert_to_ppb : bool
```

zmoon commented on monetio/models/raqms.py:
```diff
@@ -30,17 +32,37 @@ def open_dataset(fname, *, surf_only=False):
-
+    convert_to_ppb : boolean
+        If true the units of the gas species will be converted to ppbv
+    var_list: list
```
```suggestion
    var_list : list of str, optional
```

zmoon commented on monetio/models/raqms.py:
```diff
@@ -52,10 +74,27 @@ def open_mfdataset(fname, *, surf_only=False):
+        # Calculate temperature from potential temperature
+        k = 0.28571428571428564 # metpy.constants kappa
+        ds['temperature_k'] = ds['ttheta']*(ds['pres_pa_mid']/100000)**k
+        ds['temperature_k'].attrs['units'] = 'K'
```
Do you want to add `long_name` as well?

zmoon replied:
@mbruckner-work if you don't want this, let me know and/or mark resolved and I'll approve the PR.

zmoon commented on monetio/models/raqms.py:
```diff
@@ -52,10 +74,27 @@ def open_mfdataset(fname, *, surf_only=False):
-    ds = xr.open_mfdataset(names, concat_dim="time", drop_variables=["theta"], combine="nested")
+    if var_list is not None:
+        var_list.extend(['lat','lon',"IDATE", "Times","psfc","delp",'pdash','ttheta'])
+        ds = xr.open_mfdataset(names, concat_dim="time", drop_variables=["theta"], combine="nested")[var_list]
+    else:
+        ds = xr.open_mfdataset(names, concat_dim="time", drop_variables=["theta"], combine="nested")
```
Most of these two lines are the same, so you could move it above the `if` and just do `ds = ds[var_list]` inside the `if`.

zmoon commented on monetio/models/raqms.py:
```diff
@@ -30,17 +32,37 @@ def open_dataset(fname, *, surf_only=False):
     ds = xr.open_dataset(names[0], drop_variables=["theta"])
     ds = _fix(ds, surf_only=surf_only)
 
+    # convert all gas species to ppbv
```
You could move this to `_fix` or a `_fix_units` called by `_fix()` to you don't have to duplicate it?

zmoon replied:
```suggestion
```
I think your code is self-documenting enough, don't need this comment

zmoon commented on monetio/models/raqms.py:
```diff
@@ -9,14 +9,16 @@
 
 
-def open_dataset(fname, *, surf_only=False):
+def open_dataset(fname, convert_to_ppb=True, surf_only=False, **kwargs):
```
Currently `kwargs` are unused. Did you want to pass them to something?
```suggestion
def open_dataset(fname, convert_to_ppb=True, surf_only=False):
```

zmoon commented on monetio/models/raqms.py:
```diff
@@ -30,17 +32,37 @@ def open_dataset(fname, *, surf_only=False):
+    convert_to_ppb : boolean
+        If true the units of the gas species will be converted to ppbv
+    var_list: list
+        List of variables to include in output. MELODIES-MONET should only read in
+        variables need to plot in order to save on memory and simulation cost
+        especially for vertical data. If None, will read in all model data.
```
```suggestion
        List of variables to include in output. MELODIES MONET should only read in
        variables needed to plot in order to save on memory and simulation cost
        especially for vertical data. If ``None`` (default), will read in all model data.
```

zmoon commented on monetio/models/raqms.py:
```diff
@@ -52,10 +74,27 @@ def open_mfdataset(fname, *, surf_only=False):
+                    ds[i].attrs["units"] = "ppbv"
+    
+    if 'ttheta' in ds.keys():
+        # Calculate temperature from potential temperature
```
Should `open_dataset` have this too?

mbruckner-work replied:
Moved into _fix, along with the unit conversion. Should now be applied to both.

zmoon commented on monetio/models/raqms.py:
```diff
@@ -52,10 +74,27 @@ def open_mfdataset(fname, *, surf_only=False):
+    
+    if 'ttheta' in ds.keys():
+        # Calculate temperature from potential temperature
+        k = 0.28571428571428564 # metpy.constants kappa
```
```suggestion
        k = 0.28571428571428564  # R/cp = kappa (unitless; value for dry air from metpy.constants)
```

zmoon reviewed (commented):
Added a few comments (and the format check has a few too). And would you mind updating the tests, at least to check for the unit conversion?

mbruckner-work committed:
Merge branch 'raqms_update' of https://github.com/mbruckner-work/monetio into raqms_update

mbruckner-work committed:
re-add changes based on zmoon comments

mbruckner-work committed:
Merge branch 'develop' into raqms_update

mbruckner-work committed:
tropomi reader shouldn't  have had changes

I don't know why there were changes. Having some bleed over on a machine between branches.

mbruckner-work committed:
updated raqms test to include units and temperature

zmoon commented on monetio/models/raqms.py:
```diff
@@ -9,13 +9,15 @@
 
 
-def open_dataset(fname, *, surf_only=False):
+def open_dataset(fname, convert_to_ppb=True, surf_only=False):
```
```suggestion
def open_dataset(fname, *, convert_to_ppb=True, surf_only=False):
```
I think it makes sense for everything except `fname` to be keyword-only

zmoon commented on monetio/models/raqms.py:
```diff
@@ -28,18 +30,24 @@ def open_dataset(fname, *, surf_only=False):
 
 
-def open_mfdataset(fname, *, surf_only=False):
+def open_mfdataset(fname, convert_to_ppb=True, var_list=None, surf_only=False):
```
```suggestion
def open_mfdataset(fname, *, convert_to_ppb=True, var_list=None, surf_only=False):
```

zmoon commented on monetio/models/raqms.py:
```diff
@@ -28,18 +30,24 @@ def open_dataset(fname, *, surf_only=False):
+        If true the units of the gas species will be converted to ppbv
+    var_list : list of str, optional
+        List of variables to include in output. MELODIES MONET should only read in
+        variables need to plot in order to save on memory and simulation cost
```
```suggestion
        variables needed to plot in order to save on memory and simulation cost
```

mbruckner-work committed:
Update monetio/models/raqms.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

mbruckner-work committed:
Update monetio/models/raqms.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

mbruckner-work committed:
Update monetio/models/raqms.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

zmoon commented on monetio/models/raqms.py:
```diff
@@ -52,21 +60,36 @@ def open_mfdataset(fname, *, surf_only=False):
+        for i in ds.variables:
+            if "units" in ds[i].attrs:
+                if ds[i].attrs["units"] == "ppv":
+                    ds[i] *= 1e9
```
This seems to be causing an issue. In CI got `FAILED tests/test_raqms.py::test_surf_only[open_dataset] - ValueError: output array is read-only` for this line.

We could try it this way instead:
```suggestion
                    with xr.set_options(keep_attrs=True):
                        ds[i] = ds[i] * 1e9
```

mbruckner-work replied:
Good to know that about CI. Will keep that in mind for future commits.

zmoon replied:
It worked the original way on my machine. And in that run in CI it worked in Python 3.9 but failed in 3.10 and 3.11. I looked a bit at the env lists in the runs but nothing jumped out at me so far as to why that would happen. Seems like it would be related to `numpy` or `netcdf4` or `xarray` version but those were the same in 3.9 and 3.10.

mbruckner-work committed:
small change to in-place multiplication

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

zmoon reviewed (approved):
Thanks @mbruckner-work !
